### ISSUE TITLE
Improvement of datastructure

### ISSUE THREAD

**Issue by boersmamarcel** [AI4FinTech/FinTorch#116]
**Preamble: Task Objective**

The primary objective is to refactor the time series data handling in FinTorch. This involves:
1.  Modifying the `TimeSeriesDataset` base class and its implementations to return data in a new, standardized dictionary format.
2.  Updating all models that consume data from `TimeSeriesDataset` (e.g., TFT, CausalFormer) to work with this new data format.
3.  Ensuring the continued correct functionality of the system through comprehensive testing.

**The New Standardized Data Structure (per sample from `Dataset.__getitem__`)**

Each call to `Dataset.__getitem__` should return a single dictionary containing PyTorch tensors with the following keys and conceptual shapes. The `series_dim` dimension is crucial for multi-series support and should be present where applicable.

```python
{
    # --- Core Historical Data ---
    "past_target": torch.Tensor,  # Shape: (past_time_steps, series_dim, num_target_features)
    "past_covariates_known_future": torch.Tensor, # Shape: (past_time_steps, series_dim, num_known_future_cov_features)
    "past_covariates_unknown_future": torch.Tensor, # Shape: (past_time_steps, series_dim, num_unknown_future_cov_features)

    # --- Core Future Data (for decoders/targets) ---
    "future_covariates_known": torch.Tensor, # Shape: (future_time_steps, series_dim, num_known_future_cov_features)
    "output_target": torch.Tensor, # Shape: (future_time_steps, series_dim, num_target_features) # This is the label

    # --- Static Features (per series) ---
    "static_features_real": torch.Tensor, # Shape: (series_dim, num_static_real_features)
    "static_features_categorical": torch.Tensor, # Shape: (series_dim, num_static_categorical_features), dtype=torch.long

    # --- METADATA (Crucial for static categoricals; typically a dataset property or passed at model init) ---
    # "static_categorical_cardinalities": List[int], # Example:  for 3 categorical features

    # --- OPTIONAL ENHANCEMENTS (Implement if feasible and models can leverage them) ---
    "past_time_features": torch.Tensor, # Shape: (past_time_steps, num_time_features) OR (past_time_steps, series_dim, num_time_features)
    "future_time_features": torch.Tensor, # Shape: (future_time_steps, num_time_features) OR (future_time_steps, series_dim, num_time_features)
    "past_target_mask": torch.Tensor, # Shape: (past_time_steps, series_dim, num_target_features) or (past_time_steps, series_dim)
    # Add other masks (past_covariates_mask, future_covariates_known_mask) if implementing masking
    # "encoder_lengths": torch.Tensor, # Shape: (series_dim,) or scalar. Actual length of each past sequence.
    # "decoder_lengths": torch.Tensor, # Shape: (series_dim,) or scalar. Actual length of each future sequence.
}
```
*   `num_target_features`: Often 1, but can be >1 if the target for each series is itself multivariate.
*   When batched by `DataLoader`, an initial `batch_size` dimension will be added to all tensors. For example, `past_target` will become `(batch_size, past_time_steps, series_dim, num_target_features)`. Static features will become `(batch_size, series_dim, num_static_X_features)`.

---

**Phase 1: Dataset and DataLoader Updates**

**Step 1.1: Modify `TimeSeriesDataset` Base Class**
*   **File:** `fintorch/datasets/base/base_dataset.py`
*   **Action:**
    1.  Change the return type annotation of the `__getitem__` method from `Tuple, Dict, Dict, torch.Tensor]` to `Dict`.
    2.  Review the abstract properties (e.g., `time_steps`, `future_steps`, `series_dim`, `features_dim`). The concept of a single `features_dim` might become ambiguous. Consider if new abstract properties are needed to describe the dimensions of the new granular features (e.g., `num_known_future_covariates`, `num_static_real_features`). For now, ensure existing properties remain meaningful or document how they relate to the new structure.
    3.  Add a new abstract property (or a concrete method if derivable) for `static_categorical_cardinalities`:
        ```python
        @property
        @abstractmethod
        def static_categorical_cardinalities(self) -> List[int]:
            """Returns a list of cardinalities for each static categorical feature."""
            pass
        ```

**Step 1.2: Refactor `SimpleSyntheticDataset`**
*   **File:** `fintorch/datasets/synthetic/simpleSynthetic.py`
*   **Action:**
    1.  **Update `__init__`:**
        *   The current `features_dim` parameter for synthetic generation needs to be re-evaluated. Decide how this single dimension will be split into:
            *   `num_target_features` (for `past_target` and `output_target`)
            *   `num_known_future_cov_features` (for `past_covariates_known_future` and `future_covariates_known`)
            *   `num_unknown_future_cov_features` (for `past_covariates_unknown_future`)
        *   You might need to add new parameters to `__init__` to specify these counts (e.g., `num_target_features=1`, `num_known_cov_features=X`, `num_unknown_cov_features=Y`). Document this decision.
        *   Similarly, for static features, decide how the synthetic generation maps to `num_static_real_features` and `num_static_categorical_features`. Add parameters if needed.
    2.  **Implement `static_categorical_cardinalities` property:** This should return a list of integers representing the number of unique categories for each synthetically generated static categorical feature.
    3.  **Modify `__getitem__`:**
        *   Generate synthetic data according to the new feature separation.
        *   Populate and return a single dictionary matching the new standardized structure.
        *   Ensure `past_target` is correctly extracted/generated.
        *   Ensure `output_target` (your current `target`) is correctly assigned.
        *   Split generated static data into `static_features_real` and `static_features_categorical` (ensure categorical are long integers).
        *   **Optional:** Implement generation for `past_time_features`, `future_time_features`, and masks if you choose to include these optional enhancements.
    4.  **Dimensionality:** Pay close attention to the `series_dim`. All relevant tensors should have this dimension correctly. For static features, the shape from `__getitem__` should be `(series_dim, num_static_X_features)`.

**Step 1.3: Refactor `ElectricityDataset`**
*   **File:** `fintorch/datasets/electricity_simple.py`
*   **Action:**
    1.  **Inheritance:** Ensure `ElectricityDataset` inherits from the updated `TimeSeriesDataset` base class and implements all its abstract methods/properties, including `static_categorical_cardinalities`. If there are no static categorical features, this property can return an empty list.
    2.  **Modify `__getitem__`:**
        *   Transform the raw electricity data into the new standardized dictionary format.
        *   **Feature Identification:** This is crucial. You need to identify which columns in the electricity data correspond to:
            *   `past_target` / `output_target` (e.g., electricity consumption).
            *   `past_covariates_known_future` / `future_covariates_known` (e.g., date/time features like day of week, hour, holidays, if their future values are known).
            *   `past_covariates_unknown_future` (e.g., weather data from the past whose future is not used/known).
            *   `static_features_real` / `static_features_categorical` (e.g., client ID, region ID if these are static per series).
        *   If the dataset provides timestamps, generate `past_time_features` and `future_time_features`.
        *   Handle `series_dim`. If the original data is single-series but your framework expects a `series_dim`, ensure it's shaped as `(time_steps, 1, features_dim)`.
    3.  **Data Scaling:** Ensure existing scaling logic is applied correctly to the new, separated feature tensors.

**Step 1.4: Refactor Other Datasets (If Any)**
*   **Action:** If there are other dataset classes inheriting from `TimeSeriesDataset`, apply the same refactoring logic as in Step 1.2 and 1.3.

**Step 1.5: Verify DataLoader Compatibility**
*   **Action:**
    1.  The standard PyTorch `DataLoader` should collate batches correctly if `__getitem__` returns a dictionary of tensors where corresponding tensors across samples have compatible shapes.
    2.  If you have any custom `collate_fn` functions associated with these datasets, review and update them to handle the new single dictionary format.

---

**Phase 2: Model Updates and Verification**

**Step 2.1: Refactor `TemporalFusionTransformerModule`**
*   **File(s):** Likely in `fintorch/models/timeseries/tft/` (both the core model and the Lightning `TemporalFusionTransformerModule`).
*   **Action:**
    1.  **Update `__init__` (Lightning Module):**
        *   The parameters `past_inputs`, `future_inputs`, `static_inputs` (which currently expect dictionaries like `{"past_data": 1}`) will need to change. They should now reflect the number of features expected from the new specific keys. For example:
            *   `num_past_target_features: int`
            *   `num_past_known_cov_features: int`
            *   `num_past_unknown_cov_features: int`
            *   `num_future_known_cov_features: int`
            *   `num_static_real_features: int`
            *   `num_static_categorical_features: int`
            *   `static_categorical_cardinalities: List[int]` (This is crucial for setting up embedding layers).
        *   Update the model's internal configuration based on these new parameters.
    2.  **Update `forward` method (Core Model and Lightning Module's step methods):**
        *   The `forward` method (and `training_step`, `validation_step`, `test_step` in the Lightning module) will now receive a single dictionary `batch`.
        *   Access tensors using their new keys: `batch["past_target"]`, `batch.get("past_covariates_known_future")` (use `.get()` for optional fields like time features or masks if you implement them).
        *   Map these tensors to the TFT architecture's specific input requirements (e.g., historical known/unknown inputs, static inputs, future known inputs).
    3.  **Series Handling:** The `series_selection_method` and `series_aggregation` logic needs to be checked. It should operate on the input tensors which now explicitly carry the `series_dim`. For example, if `past_target` has shape `(batch_size, past_time_steps, series_dim, num_target_features)`, aggregation or selection will happen along the `series_dim`.

**Step 2.2: Refactor `CausalFormer`**
*   **File(s):** Implementation likely in `fintorch/models/timeseries/causalformer/` and tests in `tests/models/causalformer/`.
*   **Action:**
    1.  **Update `__init__` (if it has a Lightning Wrapper or takes feature counts):** Similar to TFT, adjust parameters to expect counts for the new granular feature types and `static_categorical_cardinalities` if it uses static categorical features.
    2.  **Update `forward` method (and step methods if applicable):**
        *   Modify to accept the single `batch` dictionary.
        *   Access input tensors using the new keys.
        *   Ensure the data is correctly fed into CausalFormer's specific layers (e.g., causal convolution, multi-head attention).

**Step 2.3: Refactor Other Models (If Any)**
*   **Action:** If other models consume data from `TimeSeriesDataset`, apply similar refactoring logic to their input handling and configuration.

---

**Phase 3: Comprehensive Testing and Verification**

**Step 3.1: Dataset Unit Tests**
*   **Action:** For each refactored dataset (`SimpleSyntheticDataset`, `ElectricityDataset`, etc.):
    1.  Write new unit tests or update existing ones.
    2.  Verify that `__getitem__(idx)` returns a dictionary with all the expected keys from the new data structure.
    3.  For each key, check that the corresponding tensor has the correct shape (including `series_dim`) and `dtype`.
    4.  Test with `series_dim=1` and `series_dim > 1` if applicable.
    5.  Verify the `static_categorical_cardinalities` property returns correct values.

**Step 3.2: Model Unit Tests**
*   **Action:** For each refactored model (`TemporalFusionTransformerModule`, `CausalFormer`, etc.):
    1.  Write new unit tests or update existing ones.
    2.  Create a mock batch of data in the new dictionary format (or use a sample from a refactored dataset).
    3.  Pass this mock batch to the model's `forward` method (and to Lightning module's step methods).
    4.  Verify that the model can process the input without errors and produces an output tensor of the expected shape.
    5.  Test with different configurations of the model (e.g., different `series_selection_method` for TFT).

**Step 3.3: Integration Tests (Training Runs)**
*   **Action:**
    1.  Update existing integration tests that involve training a model on a dataset.
    2.  Ensure you can initialize a refactored dataset, wrap it in a `DataLoader`, initialize a refactored model, and run `trainer.fit(...)` without errors.
    3.  Example to update (from your documentation):
        ```python
        # Ensure SimpleSyntheticDataset and TemporalFusionTransformerModule are the refactored versions
        synthetic_data = SimpleSyntheticDataset(...)
        tft_model = TemporalFusionTransformerModule(...) # Pass new config params
        trainer.fit(tft_model, DataLoader(synthetic_data))

        # Ensure CausalFormer is the refactored version
        causalformer_model = CausalFormer(...) # Pass new config params
        trainer.fit(causalformer_model, DataLoader(synthetic_data))
        ```
    4.  Test with both `SimpleSyntheticDataset` and `ElectricityDataset`.
    5.  If possible, check if the loss decreases as a basic sanity check.

**Step 3.4: Review Existing Test Suites**
*   **Action:**
    1.  Thoroughly review any existing test suites, especially those in `tests/models/causalformer/` and any related to TFT or datasets.
    2.  Update these tests to reflect the new data structures and model input APIs.
    3.  Ensure all existing relevant tests pass after the refactoring.

---

**General Guidelines for the AI Agent:**

*   **Iterative Approach:** Refactor one dataset, then one model, and test thoroughly before moving to the next.
*   **Version Control:** Commit changes frequently with clear messages.
*   **Documentation:**
    *   Document any assumptions made, especially regarding the splitting of `features_dim` from the old structure into new granular features.
    *   Update comments and docstrings in the code to reflect the new data structures and API changes.
    *   Update any relevant Markdown documentation in the FinTorch repository that describes the data format.
*   **Clarity on `series_dim`:** Ensure the `series_dim` is handled consistently. Tensors from `__getitem__` should have it, and models should be aware of it.
*   **Optional Fields:** When accessing optional fields from the batch dictionary (like time features or masks, if implemented), use `batch.get("optional_key", default_value)` to avoid `KeyError` if a dataset doesn't provide them. Models should be able to function with or without these optional inputs if designed so.
*   **Backward Compatibility (Consideration):** This refactoring is a breaking change. If backward compatibility is a major concern (e.g., for users of a released version of FinTorch), this would require a more complex strategy (e.g., versioning datasets/models or providing adapter layers), which is outside the scope of these current instructions. Assume this is a forward-looking refactor for an unreleased/development version.
